libs/@blockprotocol/graph/src/codegen/postprocess/generate-link-and-target-definitions.ts

```diff
@@ -0,0 +1,237 @@
+import { VersionedUrl } from "@blockprotocol/type-system/slim";
+
+import { mustBeDefined } from "../../shared/util/must-be-defined.js";
+import {
+  typedEntries,
+  typedValues,
+} from "../../shared/util/typed-object-iter.js";
+import { PostprocessContext } from "../context/postprocess.js";
+import { entityDefinitionNameForEntityType } from "../shared";
+
+type IdentifierToCompiledDefinition = {
+  identifier: string;
+  compiledContents: string;
+  dependentOnIdentifiers: string[];
+};
+
+/**
+ * Create the individual definition of a `LinkAndTargetEntity` for a given outgoing link from a source
+ *
+ * @param sourceName
+ * @param linkIdentifier
+ * @param targetIdentifiers
+ */
+const individualOutgoingLinkAndTargetDefinition = (
+  sourceName: string,
+  linkIdentifier: string,
+  targetIdentifiers: [string, ...string[]],
+): IdentifierToCompiledDefinition => {
+  const identifier = `${sourceName}${linkIdentifier}Links`;
+
+  const targetUnion = targetIdentifiers.join(" | ");
+  const compiledContents = `export type ${identifier} = { linkEntity: ${linkIdentifier}; target: ${targetUnion} }`;
+  return {
+    identifier,
+    compiledContents,
+    dependentOnIdentifiers: [linkIdentifier, ...targetIdentifiers],
+  };
+};
+
+/**
+ * Create a lookup map of a link entity type ID to its associated link and target definition
+ *
+ * @param entityName
+ * @param linkEntityTypeIdToLinkAndTargetIdentifiers
+ */
+const lookupDefinition = (
+  entityName: string,
+  linkEntityTypeIdToLinkAndTargetIdentifiers: Record<VersionedUrl, string>,
+): IdentifierToCompiledDefinition => {
+  const identifier = `${entityName}OutgoingLinksByLinkEntityTypeId`;
+
+  const lookupItems = typedEntries(linkEntityTypeIdToLinkAndTargetIdentifiers)
+    .map(([linkEntityTypeId, linkAndTargetIdentifier]) => {
+      return `"${linkEntityTypeId}": ${linkAndTargetIdentifier}`;
+    })
+    .join(", ");
+
+  const compiledContents = `export type ${identifier} = { ${lookupItems} }`;
+
+  return {
+    identifier,
+    compiledContents,
+    dependentOnIdentifiers: typedValues(
+      linkEntityTypeIdToLinkAndTargetIdentifiers,
+    ),
+  };
+};
+
+/**
+ * Generates the union of all outgoing link and target definitions for a given entity
+ *
+ * @param entityName
+ * @param linkAndTargetIdentifiers
+ */
+const linkAndTargetsUnionDefinition = (
+  entityName: string,
+  linkAndTargetIdentifiers: string[],
+): IdentifierToCompiledDefinition => {
+  const identifier = `${entityName}OutgoingLinksAndTargets`;
```

Comment:
As this is a union of single items, I would name it in the singular. This type describes a single outgoing link and target (which can be satisfied in multiple different ways).

Then you don't get oddities likes `FooOutgoingLinksAndTargets[]` which looks like an array of arrays but is an array of single link/target objects.

```suggestion
  const identifier = `${entityName}OutgoingLinkAndTarget`;
```

Reply:
I had named it that way before, and then found it confusing in usage and swapped it around for some reason.. I wish I knew why.

Regardless, definitely agree with the general convention on union namings:
[Rename OutgoingLinksAndTargets union to OutgoingLinkAndTarget](https://github.com/blockprotocol/blockprotocol/pull/1158/commits/9f0576eb32a34230090296de4daf0fa0d82c6625)